File: src/bairstow-pyrs/robin.rs
use std::collections::HashMap;
use std::*;

struct SlNode {
    next: ST0,
    data: ST1,
}

impl SlNode {
    fn __init__<T0>(&self, data: T0) {
        // "initialization
        //
        //         Keyword Arguments:
        //             data (type):  description
        //         ";
        self.next = self;
        self.data = data;
    }
}
struct Robin {
    cycle: ST0,
}

impl Robin {
    // "Round Robin
    //
    //     Raises:
    //         StopIteration:  description
    //
    //     Returns:
    //         dtype:  description
    //     ";
    const __slots__: _ = "cycle";
    fn __init__(&self, num_parts: i32) {
        self.cycle = (0..num_parts)
            .iter()
            .map(|k| SlNode(k))
            .collect::<Vec<_>>()
            .collect::<Vec<_>>();
        let mut sl2 = self.cycle[-1];
        for sl1 in self.cycle {
            sl2.next = sl1;
            sl2 = sl1;
        }
    }
    fn exclude<RT>(&self, from_part: i32) -> RT {
        // "iterator
        //
        //         Returns:
        //             robin_iterator
        //         ";
        return robin_iterator(self, from_part);
    }
}
struct robin_iterator {
    cur: ST0,
}

impl robin_iterator {
    fn __init__<T0>(&self, Robin: T0, from_part: i32) {
        // "[summary]
        //
        // Arguments:
        //     Robin (type):  description
        // ";
        self.cur = Robin::cycle[from_part];
    }
    fn __iter__<RT>(&self) -> RT {
        // "iterable
        //
        // Returns:
        //     robin_iterator:  itself
        // ";
        return self;
    }
    fn next<RT>(&self) -> RT {
        // "next
        //
        // Raises:
        //     StopIteration:  description
        //
        // Returns:
        //     robinink:  description
        // ";
        self.cur = self.cur.next;
        if self.cur != self.stop {
            return self.cur.data;
        } else {
            raise!(StopIteration); //unsupported
        }
    }
    fn __next__<RT>(&self) -> RT {
        // "[summary]
        //
        // Returns:
        //     dtype:  description
        // ";
        return self.next();
    }
}
fn main() {
    let r = Robin(5);
    for k in r.exclude(3) {
        println!("{:?} ", k);
    }
}

File: src/bairstow/robin.py
from typing import List


class SlNode:
    next: "SlNode"
    data: int

    def __init__(self, data: int):
        """_summary_

        Args:
            data (int): _description_
        """
        self.next = self
        self.data = data


class RobinIterator:
    __slots__ = ("cur", "stop")
    cur: SlNode
    stop: SlNode

    def __init__(self, node: SlNode) -> None:
        """_summary_

        Args:
            node (SlNode): _description_
        """
        self.cur = self.stop = node

    def __iter__(self) -> "RobinIterator":
        """_summary_

        Returns:
            RobinIterator: _description_
        """
        return self

    def next(self) -> int:
        """_summary_

        Raises:
            StopIteration: _description_

        Returns:
            int: _description_
        """
        self.cur = self.cur.next
        if self.cur != self.stop:
            return self.cur.data
        else:
            raise StopIteration()

    def __next__(self):
        """_summary_

        Returns:
            _type_: _description_
        """
        return self.next()


class Robin:
    __slots__ = "cycle"
    cycle: List[SlNode]

    def __init__(self, num_parts: int):
        """_summary_

        Args:
            num_parts (int): _description_
        """
        self.cycle = list(SlNode(k) for k in range(num_parts))
        sl2 = self.cycle[-1]
        for sl1 in self.cycle:
            sl2.next = sl1
            sl2 = sl1

    def exclude(self, from_part: int) -> RobinIterator:
        """_summary_

        Args:
            from_part (int): _description_

        Returns:
            RobinIterator: _description_
        """
        return RobinIterator(self.cycle[from_part])


if __name__ == "__main__":
    r = Robin(5)
    for k in r.exclude(3):
        print(k)
